### PR TITLE
Add section on ".qnemo" checkpoints

### PR DESCRIPTION
**Pull request by janekl**
# What does this PR do ?

Add section on ".qnemo" checkpoints to https://github.com/NVIDIA/NeMo/pull/9329.

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [x] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: docs/source/checkpoints/intro.rst [hunk 1]
```diff
@@ -20,6 +20,26 @@ With sharded model weights, you can save and load the state of your training scr
 
 NeMo supports the distributed (sharded) checkpoint format from Megatron-Core. In Megatron-Core, it supports two backends: Zarr-based and PyTorch-based.
 
+Quantized Checkpoints
+---------------------
+
+NeMo also offers :doc:`Post-Training Quantization <../nlp/quantization>` workflow to convert regular ``.nemo`` models into a `TensorRT-LLM checkpoint <https://nvidia.github.io/TensorRT-LLM/architecture/checkpoint.html>`_ conventionally referred to as ``.qnemo`` checkpoints in NeMo. Such a checkpoint can be used with `NVIDIA TensorRT-LLM library <https://nvidia.github.io/TensorRT-LLM/index.html>`_ for efficient inference.
+
+Much as in the case of ``.nemo`` checkpoints, a ``.qnemo`` checkpoint is a tar file that bundles the model configuration given in ``config.json`` file and ``rank{i}.safetensors`` files storing model weights for each rank separately. Additionally a ``tokenizer_config.yaml`` file is saved which is just ``tokenizer`` section from ``model_config.yaml`` file from the original NeMo model. This configuration file defines a tokenizer for the model given.
```
Comment (yaoyu-33) [inline docs/source/checkpoints/intro.rst:28]: .qnemo would not support distributed checkpoint format? i.e. you saved with world_size 2 and have to load with world_size 2?

File: docs/source/checkpoints/intro.rst [hunk 2]
```diff
@@ -20,6 +20,26 @@ With sharded model weights, you can save and load the state of your training scr
 
 NeMo supports the distributed (sharded) checkpoint format from Megatron-Core. In Megatron-Core, it supports two backends: Zarr-based and PyTorch-based.
```
Comment (jgerh) [inline docs/source/checkpoints/intro.rst:21]: NeMo supports the distributed (sharded) checkpoint format from Megatron Core. In Megatron Core, it supports two backends: Zarr-based and PyTorch-based.

File: docs/source/checkpoints/intro.rst [hunk 3]
```diff
@@ -20,6 +20,26 @@ With sharded model weights, you can save and load the state of your training scr
...
 
+Quantized Checkpoints
+---------------------
+
+NeMo also offers :doc:`Post-Training Quantization <../nlp/quantization>` workflow to convert regular ``.nemo`` models into a `TensorRT-LLM checkpoint <https://nvidia.github.io/TensorRT-LLM/architecture/checkpoint.html>`_ conventionally referred to as ``.qnemo`` checkpoints in NeMo. Such a checkpoint can be used with `NVIDIA TensorRT-LLM library <https://nvidia.github.io/TensorRT-LLM/index.html>`_ for efficient inference.
```
Comment (jgerh) [inline docs/source/checkpoints/intro.rst:26]: NeMo provides a Post-Training Quantization workflow that allows you to convert regular .nemo models into a TensorRT-LLM checkpoint format, commonly referred to as .qnemo checkpoints in NeMo. These .qnemo checkpoints can then be used with the NVIDIA TensorRT-LLM library for efficient inference.

File: docs/source/checkpoints/intro.rst (hunk 1 above)
Comment (jgerh) [inline docs/source/checkpoints/intro.rst:28]: A .qnemo checkpoint, similar to .nemo checkpoints, is a tar file that bundles the model configuration specified in the config.json file along with the rank{i}.safetensors files. These .safetensors files store the model weights for each rank individually. In addition, a tokenizer_config.yaml file is saved, containing only the tokenizer section from the original NeMo model_config.yaml file. This configuration file defines the tokenizer used by the given model.

File: docs/source/checkpoints/intro.rst [hunk 4]
```diff
@@ -20,6 +20,26 @@ With sharded model weights, you can save and load the state of your training scr
...
+
+Much as in the case of ``.nemo`` checkpoints, a ``.qnemo`` checkpoint is a tar file that bundles the model configuration given in ``config.json`` file and ``rank{i}.safetensors`` files storing model weights for each rank separately. Additionally a ``tokenizer_config.yaml`` file is saved which is just ``tokenizer`` section from ``model_config.yaml`` file from the original NeMo model. This configuration file defines a tokenizer for the model given.
+
+For large quantized LLM using a directory rather than a tar file is recommended. This can be controlled with ``compress`` flag on exporting quantized models in `PTQ configuration file <https://github.com/NVIDIA/NeMo/blob/main/examples/nlp/language_modeling/conf/megatron_gpt_quantization.yaml>`_.
```
Comment (jgerh) [inline docs/source/checkpoints/intro.rst:30]: When working with large quantized LLMs, it is recommended that you use a directory rather than a tar file. You can control this behavior by setting the compress flag when exporting quantized models in the PTQ configuration file.

File: docs/source/checkpoints/intro.rst [hunk 5]
```diff
@@ -20,6 +20,26 @@ With sharded model weights, you can save and load the state of your training scr
...
+
+Example contents of a quantized model to be served using two GPUs (ranks) are shown below:
```
Comment (jgerh) [inline docs/source/checkpoints/intro.rst:32]: The following example shows the contents of a quantized model intended to be served using two GPUs (ranks):

File: docs/source/checkpoints/intro.rst [hunk 6]
```diff
@@ -20,6 +20,26 @@ With sharded model weights, you can save and load the state of your training scr
...
+.. code-block:: bash
+
+    model.qnemo
+    ├── config.json
+    ├── rank0.safetensors
+    ├── rank1.safetensors
+    ├── tokenizer.model
+    └── tokenizer_config.yaml
+
 Community Checkpoint Converter
```
Comment (jgerh) [inline docs/source/checkpoints/intro.rst:43]: Edits to 45

NVIDIA provides easy-to-use tools that enable users to convert community checkpoints into the NeMo format. These tools facilitate various operations, including resuming training, Supervised Fine-tuning (SFT), Parameter Efficient Fine-Tuning (PEFT), and deployment. For detailed instructions and guidelines, please refer to the `NeMo Framework User Guide <https://docs.nvidia.com/nemo-framework/user-guide/latest/overview.html)>`__.

File: docs/source/checkpoints/intro.rst (hunk 1 above)
Comment (janekl) [reply docs/source/checkpoints/intro.rst:28]: One thing to clarify is that these `config.json` + `rank{i}.safetensors` output is a TRT-LLM checkpoint. This should not be confused by distributed checkpoint in Nemo sense.

Anyway, the feature you asked for is not available in TRT-LLM currently. So to build a TRT-LLM engine with world_size=2 one needs to calibrate/quantize model to TRT-LLM checkpoint with world_size=2 and provide this as the input to `trtllm-build` command. In other words, world_size cannnot be changed at engine build.

File: docs/source/checkpoints/intro.rst (hunk 2 above)
Comment (janekl) [reply docs/source/checkpoints/intro.rst:21]: @yaoyu-33 that would be sth for you to account for in the destination branch `yuya/add_checkpoints_section`

File: docs/source/checkpoints/intro.rst (hunk 6 above)
Comment (janekl) [reply docs/source/checkpoints/intro.rst:43]: @yaoyu-33 this is also sth for you to address here https://github.com/NVIDIA/NeMo/pull/9329, please have a look

File: docs/source/checkpoints/intro.rst (hunk 5 above)
Comment (janekl) [reply docs/source/checkpoints/intro.rst]: Thank you for your comments, I've applied necessary changes.

**Review by yaoyu-33 — APPROVED** [review]
